pkg/ipamd/datastore/data_store.go

```diff
@@ -263,79 +445,62 @@ func (ds *DataStore) DelIPv4AddressFromStore(eniID string, ipv4 string, force bo
-			return ipAddr.IP, ipAddr.DeviceNumber, nil
-		}
-		ds.log.Errorf("AssignPodIPv4Address: current IP %s is changed to IP %s for pod(name %s, namespace %s, sandbox %s)",
-			ipAddr, k8sPod.IP, k8sPod.Name, k8sPod.Namespace, k8sPod.Sandbox)
-		return "", 0, errors.New("AssignPodIPv4Address: invalid pod with multiple IP addresses")
-	}
-	return ds.assignPodIPv4AddressUnsafe(podKey, k8sPod)
-}
 
-// It returns the assigned IPv4 address, device number, error
-func (ds *DataStore) assignPodIPv4AddressUnsafe(podKey PodKey, k8sPod *k8sapi.K8SPodInfo) (ip string, deviceNumber int, err error) {
-	for _, eni := range ds.eniIPPools {
-		if (k8sPod.IP == "") && (len(eni.IPv4Addresses) == eni.AssignedIPv4Addresses) {
-			// Skip this ENI, since it has no available IP addresses
-			ds.log.Debugf("AssignPodIPv4Address: Skip ENI %s that does not have available addresses", eni.ID)
-			continue
-		}
+	for _, eni := range ds.eniPool {
 		for _, addr := range eni.IPv4Addresses {
-			if k8sPod.IP == addr.Address {
-				// After L-IPAM restart and built IP warm-pool, it needs to take the existing running pod IP out of the pool.
-				if !addr.Assigned {
-					incrementAssignedCount(ds, eni, addr)
+			if !addr.Assigned() && !addr.inCoolingPeriod() {
+				ds.assignPodIPv4AddressUnsafe(ipamKey, eni, addr)
+				if err := ds.writeBackingStoreUnsafe(); err != nil {
+					ds.log.Warnf("Failed to update backing store: %v", err)
+					// Important! Unwind assignment
+					ds.unassignPodIPv4AddressUnsafe(eni, addr)
+					return "", 0, err
 				}
-				ds.log.Infof("AssignPodIPv4Address: Reassign IP %v to pod (name %s, namespace %s)",
-					addr.Address, k8sPod.Name, k8sPod.Namespace)
-				ds.podsIP[podKey] = PodIPInfo{IP: addr.Address, DeviceNumber: eni.DeviceNumber}
-				return addr.Address, eni.DeviceNumber, nil
-			}
-			if !addr.Assigned && k8sPod.IP == "" && !addr.inCoolingPeriod() {
-				// This is triggered by a pod's Add Network command from CNI plugin
-				incrementAssignedCount(ds, eni, addr)
-				ds.log.Infof("AssignPodIPv4Address: Assign IP %v to pod (name %s, namespace %s sandbox %s)",
-					addr.Address, k8sPod.Name, k8sPod.Namespace, k8sPod.Sandbox)
-				ds.podsIP[podKey] = PodIPInfo{IP: addr.Address, DeviceNumber: eni.DeviceNumber}
+
 				return addr.Address, eni.DeviceNumber, nil
 			}
 		}
+		ds.log.Debugf("AssignPodIPv4Address: ENI %s does not have available addresses", eni.ID)
 	}
 	ds.log.Errorf("DataStore has no available IP addresses")
 	return "", 0, errors.New("assignPodIPv4AddressUnsafe: no available IP addresses")
 }
 
-func incrementAssignedCount(ds *DataStore, eni *ENIIPPool, addr *AddressInfo) {
+// It returns the assigned IPv4 address, device number
+func (ds *DataStore) assignPodIPv4AddressUnsafe(ipamKey IPAMKey, eni *ENI, addr *AddressInfo) (string, int) {
+	ds.log.Infof("AssignPodIPv4Address: Assign IP %v to sandbox %s",
+		addr.Address, ipamKey)
+
+	if addr.Assigned() {
+		panic("addr already assigned")
+	}
```

Comment:
I can:
- return an error at runtime, which (eventually) results in the pod restarting as well.
- ignore it and overwrite an existing pod ip
- ignore it and incorrectly set up the new pod
- remove the check so we don't see a `panic()` in the code, document the expected invariant, and trust that callers never invoke this function for an already assigned ipamKey
- (current version) assert the expected invariant explicitly in code, and thus know that no callers invoke this function for an already assigned ipamKey.

Which would you like?

Reply:
The safest would be to panic for sure, since we should never be in this state. It would give a strong signal that something is very wrong in the current state of the aws-node pod.